sdk/core/azure-core/src/http/challenge_based_authentication_policy.cpp

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/core/http/policies/policy.hpp"
+#include <chrono>
+#include <vector>
+
+using Azure::Core::Context;
+using namespace Azure::Core::Http;
+using namespace Azure::Core::Http::Policies;
+using namespace Azure::Core::Http::Policies::_internal;
+using namespace Azure::Core::Http::Policies::_detail;
+
+std::unique_ptr<RawResponse> ChallengeBasedAuthenticationPolicy::Send(
+    Request& request,
+    NextHttpPolicy nextPolicy,
+    Context const& context) const
+{
+  request.SetHeader("authorization", "Bearer " + m_accessToken.Token);
+  auto result = nextPolicy.Send(request, context);
+  auto const& headers = result->GetHeaders();
+  auto const& rawDataHeader = headers.find("www-authenticate");
+  // Only re-run when Unauthozied AND www-authenticate header in the response
+  if (result->GetStatusCode() == HttpStatusCode::Unauthorized && rawDataHeader != headers.end())
+  {
+    ChallengeParameters challenge(rawDataHeader->second);
+    Credentials::TokenRequestContext tokenRequestContext;
+
+    tokenRequestContext.TenantId = std::move(challenge.TenantId);
+    tokenRequestContext.AuthorizationUri = std::move(challenge.AuthorizationUri);
+    tokenRequestContext.Scopes = std::move(challenge.Scopes);
+
+    {
+      std::lock_guard<std::mutex> lock(m_accessTokenMutex);
+      m_accessToken = m_credential->GetToken(tokenRequestContext, context);
+      request.SetHeader("authorization", "Bearer " + m_accessToken.Token);
+    }
+    return nextPolicy.Send(request, context);
+  }
+
+  // Did not re-send on any other case
+  return result;
+}
+/*
+ * the raw value ( the value of the "www-authenticate" header is in the following format
+ * "Bearer authorization/authorization_uri=[value] resource/scope=[value]";
+ */
+ChallengeParameters::ChallengeParameters(std::string const& headerValue)
+{
+  if (headerValue.rfind(_detail::BearerName) == 0)
+  {
+    Schema = _detail::BearerName;
+
+    std::vector<std::string> const parts = GetParts(headerValue, _detail::SpaceSeparator);
+    for (size_t i = 1; i < parts.size(); i++)
+    {
+      ProcessFragment(parts[i]);
+    }
+  }
+}
+
+void ChallengeParameters::ProcessFragment(std::string const& fragment)
```

Comment:
Instead of using `std::sting` as input arg, we can use tow iterators (begin and end).  That way we don't need to allocate a string for each fragment we want to process.  We would just get the positions within the source string

Reply:
at the end of the day i need the strings to pass into the header and the request, it seems that we are overengineering using the position, instead of just using strings. It makes it harder to code, test, debug, trace. imho it would also be more prone to errors using indexes